### PR TITLE
Wipe LUKS Disk Encryption Key for Root Disk from RAM during Shutdown to defeat Cold Boot Attacks

### LINKED ISSUE #997: Wipe LUKS Disk Encryption Key for Root Disk from RAM during Shutdown to defeat Cold Boot Attacks
**Is your feature request related to a problem? Please describe.**

Defeat Cold Boot Attacks by wiping LUKS disk encryption during shutdown.

What is a Cold Boot Attacks? See:

* https://www.youtube.com/watch?v=JDaicPIgn9U
* https://en.wikipedia.org/wiki/Cold_boot_attack
* https://blog.f-secure.com/cold-boot-attacks/
* https://www.usenix.org/legacy/event/sec08/tech/full_papers/halderman/halderman.pdf

**Describe the solution you'd like**
Run `cryptsetup close` at end of shutdown procedure.

Quote `cryptsetup close` (previously `cryptsetup lukseClose`) man page (bold added):

> close
> Removes the existing mapping <name> and **wipes the key from kernel memory**.

Maybe `cryptsetup close` could be done during `dracut-shutdown`?

This would not wipe all secrets from RAM to defeat a cold boot attack but at least remove one of the most important secrets, the root disk LUKS encryption key.

**Describe alternatives you've considered**

Linux kernel feature:

This issue can probably not be redirected at the Linux kernel. While a generic solution `Wipe RAM to defeat Cold Boot Attacks` (https://github.com/systemd/systemd/issues/17242) probably belongs into the kernel, this does not. For the kernel to be able to wipe the memory, encrypted LUKS devices need to be properly closed first. `cryptsetup close` does that.

systemd feature:

systemd does not wipe the LUKS disk encryption key for root disk from RAM during shutdown. And as I understand systemd developer @poettering Lennart Poettering, this isn't up to systemd either. It's up to the initrd / initramfs. (https://github.com/systemd/systemd/issues/17887)

Quote myself (https://github.com/systemd/systemd/issues/17887#issuecomment-750340608):

> Avoiding all sidelines, keeping this simple, for my understanding and for the record and please correct me if I am wrong... Summary:
> 
> `"cryptsetup close" of root device during shutdown` is already implemented.

Quote systemd developer @poettering Lennart Poettering (https://github.com/systemd/systemd/issues/17887#issuecomment-751252894):

> > `"cryptsetup close" of root device during shutdown` is already implemented.
> 
> iff your initrd/distro of choice do so. For the root disk it doesn't matter what systemd does, it matters what the initrd/distro do. hence ping the maintainers of those.

### PR DESCRIPTION
**Pull request by adrelanos**
Purpose of this pull request: Receiving some early feedback if this approach looks acceptable.

* Work in progress.
* Console output are yet to be improved and documentation written.
* Untested by me. Will test soon.
* Tests I am not sure yet how this could realistically be tested.

## Changes

* Confirm in console output if encrypted mounts (root disk) is unmounted. (Because that is a pre-condition for wiping the LUKS full disk encryption key from RAM.)
* Wipe LUKS Disk Encryption Key for Root Disk from RAM during Shutdown to defeat Cold Boot Attacks.

## Checklist
- [ ] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

Fixes #997

### CODE REVIEW

**Comment by pvalena** [comment]
@lnykryn WDYT?

File: modules.d/40sdmem/wipe.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+echo "Checking for mounted disks..."
+dmsetup ls --target crypt
```
Comment (pvalena) [inline modules.d/40sdmem/wipe.sh:2]: You should probably `depend` on this properly.

File: modules.d/40sdmem/wipe.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+echo "WIPE RAM!"
+/bin/sdmem -f
```
Comment (pvalena) [inline modules.d/40sdmem/wipe.sh:4]: This as well.

**Comment by adrelanos** [comment]
I didn't update this PR yet.

Meanwhile the code has developed a lot further . It is now being tested by me in Kicksecure. The required `depend` was already added.

I would very much like if dracut was interested to merge (upstream) this functionality.

Here's the code that I have so far:

* https://github.com/Kicksecure/security-misc/tree/master/usr/lib/dracut/modules.d/40cold-boot-attack-defense
* https://github.com/Kicksecure/security-misc/blob/master/usr/lib/dracut/modules.d/40cold-boot-attack-defense/module-setup.sh
* https://github.com/Kicksecure/security-misc/blob/master/usr/lib/dracut/modules.d/40cold-boot-attack-defense/wipe-ram-needshutdown.sh
* https://github.com/Kicksecure/security-misc/blob/master/usr/lib/dracut/modules.d/40cold-boot-attack-defense/wipe-ram.sh

Design documentation:
https://www.kicksecure.com/wiki/Dev/RAM_Wipe

General user documentation on cold boot attacks:
https://www.kicksecure.com/wiki/Cold_Boot_Attack_Defense

A related issue is https://github.com/dracutdevs/dracut/issues/1862 that's why I am currently using `echo "..." > /dev/kmsg` but that surely can be resolved.

Whatever seems required to upstream, I'd be happy to modify the code to the best of my abilities and/or licensing to accommodate dracut because I very much appreciate dracut and would like this feature to be easily, widely available. (Not only in Kicksecure.)

**Comment by adrelanos** [comment]
Wiping all the RAM is different only clean LUKS unmount.

ram-wipe is now implemented as a separate package:

* https://github.com/Kicksecure/ram-wipe
* https://www.kicksecure.com/wiki/ram-wipe
* https://www.kicksecure.com/wiki/Dev/RAM_Wipe

I'd be happy if dracut was interested in this, picking this up but upstreaming seems quite difficult for me.

Effective ram-wipe however depends on https://github.com/dracutdevs/dracut/pull/2471 which will help a ton.

**Pull request closed without merging**